### PR TITLE
[PATCH API-NEXT v2] linux-gen: crypto: implement AES-XCBC-MAC and SHA384-HMAC

### PR DESCRIPTION
**Pull request by SonicwallYhe**

### CODE REVIEW

**Comment by lumag** [comment]
Next step us getting patches checkpatch-clean. Cycle `git format-patch origin/api-next..HEAD` to get your patches and `./scripts/checkpatch.pl`/`git commit`/`git rebase -i` till checkpatch returns no issues in your patches, then do a force-push to update this PR.

**Comment by Bill-Fischofer-Linaro** [comment]
Please also execute the ODP Contributor's License Agreement. You can find the individual CLA at https://www.opendataplane.org/contributor/individual/ 

If SonicWall would prefer to execute a Corporate CLA, so that anyone in the company can contribute that form can be found here: https://www.opendataplane.org/contributor/corporate/

**Comment by SonicwallYhe** [comment]
@Bill-Fischofer-Linaro I have submitted the individual CLA.

File: platform/linux-generic/odp_crypto.c [hunk 1]
```diff
@@ -36,6 +36,9 @@
 
 #define MAX_SESSIONS 32
 
+#define AES_BLOCK_SIZE 16
+typedef uint32_t aes_block[4];
```
Comment (lumag) [inline platform/linux-generic/odp_crypto.c]: Please use `uint32_t*` instead of adding new type.

File: platform/linux-generic/odp_crypto.c [hunk 2]
```diff
@@ -308,6 +319,142 @@ void packet_hmac(odp_packet_t pkt,
...
+	res[3] ^= op[3];
+}
+
+static
+odp_crypto_alg_err_t aesxcbc_gen(odp_packet_t pkt,
+			      const odp_crypto_packet_op_param_t *param,
+			      odp_crypto_generic_session_t *session)
+{
+	aes_block e = {0, 0, 0, 0};
+	uint8_t *data  = odp_packet_data(pkt);
```
Comment (lumag) [inline platform/linux-generic/odp_crypto.c]: This will not work with segmented packets.

File: platform/linux-generic/odp_crypto.c [hunk 3]
```diff
@@ -308,6 +319,142 @@ void packet_hmac(odp_packet_t pkt,
...
+	do_pad_xor((uint8_t *)e, data, len);
+	if (len == AES_BLOCK_SIZE) {
+		xor_block(e, (const uint32_t*) (session->auth.key + 16));
+	}
+	else
+	{
+		xor_block(e, (const uint32_t*) (session->auth.key + 16*2));
+	}
+	EVP_EncryptUpdate(ctx, hash_out, &dummy_len, (uint8_t *)e, sizeof(e));
+	EVP_CIPHER_CTX_free(ctx);
```
Comment (lumag) [inline platform/linux-generic/odp_crypto.c]: XCBC-MAC computation should be extracted to a separate function.

File: platform/linux-generic/odp_crypto.c [hunk 4]
```diff
@@ -308,6 +319,142 @@ void packet_hmac(odp_packet_t pkt,
...
+		{ 0x03030303, 0x03030303, 0x03030303, 0x03030303 },
+	};
+	EVP_CIPHER_CTX *ctx;
+	int dummy_len = 0;
+	
+	/* Set function */
+	if (ODP_CRYPTO_OP_ENCODE == session->p.op)
+		session->auth.func = aesxcbc_gen;
+	else
+		session->auth.func = aesxcbc_check;
```
Comment (lumag) [inline platform/linux-generic/odp_crypto.c]: You also need to set auth.init func here. For now you can use nullinit function

File: platform/linux-generic/odp_crypto.c [hunk 5]
```diff
@@ -308,6 +319,142 @@ void packet_hmac(odp_packet_t pkt,
...
+
+	ctx = EVP_CIPHER_CTX_new();
+	EVP_EncryptInit_ex(ctx, EVP_aes_128_ecb(), NULL, session->p.auth_key.data, NULL);
```
Comment (lumag) [inline platform/linux-generic/odp_crypto.c]: `EVP_aes_128_ecb()` should become a parameter here that is stored as `session->auth.evp_cipher`.

File: platform/linux-generic/odp_crypto.c (hunk 2 above)
Comment (SonicwallYhe) [reply platform/linux-generic/odp_crypto.c]: I am  not very clear about here. So which api I should use?

Comment (lumag) [reply platform/linux-generic/odp_crypto.c]: odp_packet_offset to loop through the packet.

**Comment by SonicwallYhe** [comment]
@lumag How can I operation to fix original commits?

**Comment by lumag** [comment]
The easiest way is to edit them during `git rebase -i`.

**Comment by SonicwallYhe** [comment]
@lumag Could you please tell it more detail about the operation?
I use git rebase -i but still don't clear about the operation.

**Comment by lumag** [comment]
@SonicwallYhe git rebase -i, change pick to edit, save, quit. After git rebase stops, edit the file, do git commit --amend, then git format-patch -1 && checkpatch it. If the patch satisfies you, do git rebase --continue to shift to next patch. Then, after all patches are done ,do git push -f

**Comment by Bill-Fischofer-Linaro** [comment]
@SonicwallYhe were you planning to issue a v2 of this PR to address the earlier comments / issues?

**Comment by SonicwallYhe** [comment]
@Bill-Fischofer-Linaro Yes, I am confused how should I do. I should create a new branch and start a new PR, right?

**Comment by Bill-Fischofer-Linaro** [comment]
@SonicwallYhe No, you simply revise your copy of it and when you push it back to your GitHub it will automatically become Version 2 of this PR.

**Comment by SonicwallYhe** [comment]
@lumag Do you know how to run the function crypto_suite_term() so that I can check if the change is correct or not in the new version

**Comment by lumag** [comment]
@SonicwallYhe just run crypto_main program, it will report all failures.

**Comment by lumag** [comment]
@SonicwallYhe we are moving closer towards TigerMoth release. We would like to include this into the release. Any progress on updating implementation?

**Comment by Bill-Fischofer-Linaro** [comment]
@SonicwallYhe this PR needs to be rebased to pick up the latest merge changes. Since the APIs are now part of ODP Tiger Moth RC2, this can be based against the master branch.

**Comment by lumag** [comment]
@SonicwallYhe I'll finish XCBC, if you don't mind, you can update this PR to contain only SHA384 support.

**Comment by SonicwallYhe** [comment]
@lumag Thanks for your kindly help. And I'd like to finish XCBC myself as it's an important part for me. I will   finish it soon

**Comment by lumag** [comment]
@SonicwallYhe fine with me. Ping us if you need additional help.
To rebase this PR on top of updated api-next branch please do `git rebase --onto origin/api-next origin/api-next@{1}`

**Comment by SonicwallYhe** [comment]
@lumag @Bill-Fischofer-Linaro  I push the fix and it should work, please help to check.

**Comment by lumag** [comment]
@SonicwallYhe This branch has conflicts, it contains commits from further api-next, so could you please do a clean rebase on top of api-next? `git rebase --onto origin/api-next origin/api-next@{1}` might help you. Otherwise it is nearly impossible to review your changes.

**Comment by SonicwallYhe** [comment]
@lumag The conflicting files aren't my change. I am not sure how to resolve these files.

**Comment by lumag** [comment]
@SonicwallYhe by doing interactive rebase and selecting only your changes.

**Comment by SonicwallYhe** [comment]
@lumag how can I do it as you said

**Comment by lumag** [comment]
@SonicwallYhe please see the git rebase command that I suggested to you. [Git Book](https://git-scm.com/book/en/v2) has a [nice chapter](https://git-scm.com/book/en/v2/Git-Branching-Rebasing) about rebasing.

**Comment by lumag** [comment]
@SonicwallYhe If nothing else works, I create a branch containing your changes rebased on top of latest api-next. Could you please do the following commands in your repo, after checking out this branch:

```sh
git remote add lumag https://github.com/lumag/odp
git reset --hard lumag/sonicwall-xcbc-clean
```
Then do a force-push of your branch to the repo passing `-f` argument to your regular `git push` command.

**Comment by Bill-Fischofer-Linaro** [comment]
Superseded by PR #536. @SonicwallYhe please close this PR.

**Pull request closed without merging**